zerocloud/proxyquery.py

```diff
@@ -1694,10 +1712,15 @@ def _process_response(self, conn, request):
         return self.process_server_response(conn, request, resp)
 
     def _connect_exec_node(self, obj_nodes, part, request,
-                           logger_thread_locals, cnode, request_headers):
+                           logger_thread_locals, cnode, request_headers,
+                           known_nodes, salt):
         self.app.logger.thread_locals = logger_thread_locals
         conn = None
-        for node in obj_nodes:
+        for node in chain(known_nodes, obj_nodes):
+            if (known_nodes and node in known_nodes) \
+                    or (not known_nodes and cnode.location):
```

Comment:
I think it's better to redo the logic to be more self-descriptive then.
for example:
~~~~
for node in chain([candidate_node], obj_nodes):
    if cnode.location:
        if (not candidate_node or node is candidate_node):
~~~~